web3/_utils/abi.py

```diff
@@ -12,10 +17,17 @@
 from eth_abi.codec import (
     ABICodec,
 )
+from eth_abi.grammar import (
+    ABIType,
```

Comment:
So is the idea that anything that is public API should be importable from the top-level module?  I've often avoided this pattern with library design because it seems like an unnecessary formalism to add an import in the top-level `__init__` that can potentially become stale.  On the other hand, you can forego that incantation altogether and just import things from their respective sub-modules.  Thoughts on that?  I guess I'm also a little unclear on our specific methods for distinguishing something as public API.

Reply:
I don't feel strongly about top level imports.  Do as you see fit there.

Public/Private API.

- It's a public API if it's used in 3rd party code **or** we intend for it to be used in 3rd party code.  I.E. we are using `eth_abi.grammar.parse` in the `web3.py` codebase so that API **must** be considered public.
- All public APIs **must** be documented.
- All public APIs are subject to semver style deprecation policy, meaning no breaking changes wiithout bumping the major version number, and breaking changes **should** always go through a formal deprecation cycle and raise proper warnings before being changed.

Those are my definitions/rules that I don't believe are actually written down formally anywhere.

Reply:
BTW, I'm open to suggestions to alternative definitions, etc.

Reply:
That all seems reasonable.  I think I'll probably forego the top-level imports though.

Reply:
Added documentation for `eth_abi.grammar` module in eth-abi project.